### ISSUE TITLE
Better chol and leverage score

### ISSUE THREAD

**Issue by YinTat** [ConstrainedSampler/PolytopeSamplerMatlab#23]
There are two things need to be done
1) run both col and leverage score row by row instead of col by col
2) Reorder the loop so that it is more cache friendly

**Comment by YinTat** [comment]
Currently, the program timing on my computer:
All timing is how fast it takes for single-core but runs in parallel to have a heavy load on memory.
peak is calculated using my computer, 16 flops per cycle, 4GHz.

```
   PackedCSparse      peak    matlab   chol x1   chol x4   chol x8   leve x1   leve x4   leve x8
          Recon1  7.29e-05  4.57e-03  1.10e-03  1.21e-03  1.75e-03  2.78e-03  3.21e-03  4.26e-03
          Recon2  1.97e-04  8.74e-03  2.66e-03  3.43e-03  5.05e-03  7.87e-03  9.44e-03  1.30e-02
          Recon3  1.06e-03  2.11e-02  1.43e-02  2.32e-02  3.56e-02  5.17e-02  6.83e-02  1.22e-01
          Recon4  1.47e-02  4.55e-01  3.75e-01  7.58e-01  1.23e+00  7.49e-01  1.43e+00  2.38e+00
```

For Recon3 chol, we take 3.56e-02 / 8 = 4.45e-3, matlab take 2.11e-02, peak take 1.06e-03.
So, we are 4.2 times slower than peak, Matlab is 4.7 times slower than us.

For Recon4 chol, we take 1.23e+00 / 8 = 1.53e-1, matlab take 4.55e-01, peak take 1.47e-02.
So, we are 10.4 times slower than peak, Matlab is 3 times slower than us.

Our program is probably memory-bound. Probably we need to rewrite the whole program by using a more advanced implementation of chol. This is out of scope for this project.

**Comment by YinTat** [comment]
I will keep this issue here. Maybe we can think again about this for the next version.
